Save recent searches with name attribute

szewai opened the pull request:
#### c58c088437779ba48d081ebc83242e3b71f34ca7
<pre>
Save recent searches with name attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=252139">https://bugs.webkit.org/show_bug.cgi?id=252139</a>
rdar://105369635

Reviewed by Darin Adler.

autosave attribute is non-standard and is not tracked in any active proposal, so we should stop using it to decide
whether a search item can be saved. This patch changes the logic to use name attribute to match Chrome and Firefox&apos;s
behavior.

New test: fast/forms/search/search-recent-searches.html

* LayoutTests/TestExpectations:
* LayoutTests/fast/forms/search/search-recent-searches-expected.txt: Added.
* LayoutTests/fast/forms/search/search-recent-searches.html: Added.
* LayoutTests/imported/blink/fast/forms/search-popup-crasher.html:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/platform/cocoa/SearchPopupMenuCocoa.mm:
(WebCore::saveRecentSearches):
* Source/WebCore/rendering/RenderSearchField.cpp:
(WebCore::RenderSearchField::recentSearches):
(WebCore::RenderSearchField::autosaveName const):
* Source/WebCore/rendering/RenderSearchField.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::recentSearches):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/260316@main">https://commits.webkit.org/260316@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/021957337364adcb76a61ee5079635e24e92d9db

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/107951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/17012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/40837 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/117086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/111839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/18398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/8329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/100143 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/113714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/18398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/40837 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/100143 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/18398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/40837 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/100143 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/9931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/40837 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/10638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/8329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/16083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/40837 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/12212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -202,7 +202,7 @@ Visibility RenderSearchField::visibilityForCancelButton() const
 const AtomString& RenderSearchField::autosaveName() const
 {
-    return inputElement().attributeWithoutSynchronization(autosaveAttr);
+    return inputElement().attributeWithoutSynchronization(nameAttr);
```
Can we create a test here? I don’t think that just changing this with no test coverage is a good idea.

I understand this is not a standard and unprecedented in other web browsers. But is there any app or website relying on this behavior in WebKit? We need to quickly research that before removing any feature.

darinadler replied:
Normally we require regression tests for any change to WebKit. Or we explain why this needs to be an exception to that rule.

szewai replied:
> Can we create a test here? I don’t think that just changing this with no test coverage is a good idea.

I added a test that checks the saved items via `internals.recentSearches`. It's indeed helpful. With it I found the items are not saved to disk because we didn't create parent directory of the plist file.

> I understand this is not a standard and unprecedented in other web browsers. But is there any app or website relying on this behavior in WebKit? We need to quickly research that before removing any feature.

I checked existing tests, and there's only one old imported test uses "autosave" attribute, and does not check if the item is actually saved. Based on the facts that
1. we don't have any test coverage
2. it's not documented anywhere (I checked standard and [mdn](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/search))
3. it might be broken for a while (as we didn't ensure directory exists)

The chance of app or website relying on this might be small. I think we could just go ahead and do this for compatibility, and watch out for breakage.

darinadler commented on LayoutTests/TestExpectations:
```diff
@@ -132,6 +132,7 @@ fast/scrolling/latching [ Skip ]
 
 fast/forms/search/search-padding-cancel-results-buttons.html [ Skip ]
 fast/forms/search/search-results-hidden-crash.html [ Skip ]
+fast/forms/search/search-recent-searches.html [ Skip ]
```
Does this section need a comment? Not sure it's self explanatory.

szewai replied:
I think this is because style of search field (including buttons) is different on different platforms, and these tests aim to test events on certain elements. Will add a comment.

darinadler commented on LayoutTests/imported/blink/fast/forms/search-popup-crasher.html:
```diff
@@ -17,7 +17,7 @@
 <p>This page tests that a page with a search popup does not crash, as reported in <a href='https://bugs.webkit.org/show_bug.cgi?id=37141'>this bug</a>. Below is a search input that the test automatically fills then submits. Pass if this does not crash.</p><br><br>
 <FORM action="" id="searchForm">
-<INPUT autosave="my.search" id="searchBox" type="search" results="10" value=""><br>
+<INPUT name="my.search" id="searchBox" type="search" results="10" value=""><br>
```
I'm now worried that we may basically be introducing a feature here: auto saving for the first time. Apparently it has been broken for a long time. And even when we had it, websites had to opt in and almost none did. Now we will be saving on any form that uses the `name` attribute. I guess other browsers are doing that but does the UI in Safari for clearing his true work? Does this work automatically in other WebKit contexts? Do they offer a way to clear history? Does this create fingerprinting risk?

I‘m not an expert on all of this but I think it might require a little discussion with other folks at Apple.

szewai replied:
These are good questions. Actually I have researched a bit (since I am new to this code) on these when I made this patch, and here are the answers:
> does the UI in Safari for clearing this true work?  

We have SPI to fetch/remove this type of data; internal clients like Safari should be able to fetch/remove it.

> Does this work automatically in other WebKit contexts? Do they offer a way to clear history?

We don't have API for public clients to fetch/clear it. As we will start to save this data to disk, I think we should provide a way for WebKit client to clear it. I will make a followup to promote the SPI to API.

> Does this create fingerprinting risk?

This should not create fingerprinting risk because it is not available for web client (cannot be queried with Web API).
But we do have a problem that we don't separate this data by WebsiteDataStore (user might see the same search items when they load page in different sessions), and I made a patch at https://github.com/WebKit/WebKit/pull/9948 to fix this.

darinadler replied:
It may not be enough to deal with just the WebKit project part of this, making sure that clients _could_ offer a way to clear this. I think we might to design WebKit API so that clients let us know they have a way to clear it and only then do we save it. Having hidden data that is a history of what you did on a site is a potential privacy problem if that old data shows up later.

geoffreygaren replied:
One way to do this would be to offer a 'savesSearchElementHistory' or 'shouldSaveSearchElementHistory' property on WKPreferences, which defaulted to false.

A webpage can already trigger a similar effect with [a few lines of LocalStorage code](https://stackoverflow.com/questions/63546745/save-input-search-history). But maybe triggering the effect implicitly and in a new circumstance warrants opt-in logic.

szewai replied:
> I think we might to design WebKit API so that clients let us know they have a way to clear it and only then do we save it.

Generally, when a data type (normally comes with a new feature) is introduced, the WebKit clients might not realize in time that they need to turn on a new flag, or they might not check whether the data type can be cleared before turning on the flag (likely not all clients care about privacy as we do).
So I tend to think it should be WebKit's responsibility to ensure there is a way to clear the data before introducing a new type, as Geoff mentioned in https://github.com/WebKit/WebKit/pull/10104#pullrequestreview-1298367093.

> One way to do this would be to offer a 'savesSearchElementHistory' or 'shouldSaveSearchElementHistory' property on WKPreferences, which defaulted to false.

For search field history specifically, I think there are two cases:
1. For WebKit clients who have history saved on disk (as I am not sure when the feature is broken and assume it works when first introduced), they are already in the situation of hidden data (cause no API to clear it), so it may not help if we turn off the feature in new builds.
2. For WebKit clients who have no history saved on disk due to "directory is not created" issue, they will start to have data saved on disk, and I have promoted the SPI to API in https://github.com/WebKit/WebKit/pull/10104, so the client should be able to clear it in new builds.

So I think `shouldSaveSearchElementHistory` might not help much for the privacy issue; it may be more useful if we know some WebKit clients just don't want the feature at all.

> A webpage can already trigger a similar effect with [a few lines of LocalStorage code](https://stackoverflow.com/questions/63546745/save-input-search-history). But maybe triggering the effect implicitly and in a new circumstance warrants opt-in logic.

It's true similar effect could be trigger with `LocalStorage` or other type of storage (and I think google does this). But I think the search field history is more of a user-facing feature as the data is shared across sites. For example, if you type 'iPhone 14' in apple.com, 'iPhone 14' may show up in history of search field on iCloud.com (not visible to iCloud.com).

darinadler commented on Source/WebCore/rendering/RenderSearchField.h:
```diff
@@ -44,6 +44,7 @@ class RenderSearchField final : public RenderTextControlSingleLine, private Popu
     bool popupIsVisible() const { return m_searchPopupIsVisible; }
     void showPopup();
     void hidePopup();
+    Vector<RecentSearch> recentSearches() const { return m_recentSearches; }
```
This is unnecessarily inefficient because it copies the vector every time it’s called. Could return a `Span` since the caller just needs to iterate, not copy.

I guess it’s just for testing but even so I would prefer a more efficient idiom.

szewai replied:
Sure, I will change.

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -2228,6 +2229,23 @@ auto Internals::lastAutoFillButtonType(const HTMLInputElement& element) -> AutoF
 
+Vector<String> Internals::recentSearches(const HTMLInputElement& element)
+{
+    if (!element.isSearchField())
```
Seems like we should check the type of the renderer rather than the type of the element.

szewai replied:
Sure, will change.

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -2228,6 +2229,23 @@ auto Internals::lastAutoFillButtonType(const HTMLInputElement& element) -> AutoF
+    if (!element.isSearchField())
+        return { };
+
+    auto* renderer = element.renderer();
```
Typically we have to do a “layout update” before accessing renderers. There should be lots of examples of that in this file.

szewai replied:
Got it; will look.

szewai committed:
Save recent searches with name attribute
https://bugs.webkit.org/show_bug.cgi?id=252139
rdar://105369635

Reviewed by Darin Adler.

autosave attribute is non-standard and is not tracked in any active proposal, so we should stop using it to decide
whether a search item can be saved. This patch changes the logic to use name attribute to match Chrome and Firefox's
behavior.

New test: fast/forms/search/search-recent-searches.html

* LayoutTests/TestExpectations:
* LayoutTests/fast/forms/search/search-recent-searches-expected.txt: Added.
* LayoutTests/fast/forms/search/search-recent-searches.html: Added.
* LayoutTests/imported/blink/fast/forms/search-popup-crasher.html:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/platform/cocoa/SearchPopupMenuCocoa.mm:
(WebCore::saveRecentSearches):
* Source/WebCore/rendering/RenderSearchField.cpp:
(WebCore::RenderSearchField::recentSearches):
(WebCore::RenderSearchField::autosaveName const):
* Source/WebCore/rendering/RenderSearchField.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::recentSearches):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/260316@main
